conduit/pipeline.go

```diff
@@ -0,0 +1,299 @@
+func MakePipelineConfig(logger *log.Logger, cfg *Config) (*PipelineConfig, error) {
+	if cfg == nil {
+		return nil, fmt.Errorf("MakePipelineConfig(): empty conduit config")
+	}
+
+	// double check that it is valid
+	if err := cfg.Valid(); err != nil {
+		return nil, fmt.Errorf("MakePipelineConfig(): %w", err)
+	}
+
+	pCfg := PipelineConfig{pipelineLogLevel: logger.Level.String(), ConduitConfig: cfg}
+
+	// Search for pipeline configuration in data directory
+	autoloadParamConfigPath := filepath.Join(cfg.ConduitDataDir, autoLoadParameterConfigName)
+
+	_, err := os.Stat(autoloadParamConfigPath)
+	paramConfigFound := err == nil
+
+	if !paramConfigFound {
+		return nil, fmt.Errorf("MakePipelineConfig(): could not find %s in data directory (%s)", autoLoadParameterConfigName, cfg.ConduitDataDir)
+	}
+
+	logger.Infof("Auto-loading Conduit Configuration: %s", autoloadParamConfigPath)
+
+	file, err := os.Open(autoloadParamConfigPath)
+
+	err = viper.ReadConfig(file)
+	if err != nil {
+		return nil, fmt.Errorf("MakePipelineConfig(): reading config error: %w", err)
+	}
+
+	err = viper.Unmarshal(&pCfg)
+	if err != nil {
+		return nil, fmt.Errorf("MakePipelineConfig(): config file (%s) was mal-formed yaml: %w", autoloadParamConfigPath, err)
+	}
+
+	if err := pCfg.Valid(); err != nil {
+		return nil, fmt.Errorf("MakePipelineConfig(): config file (%s) had mal-formed schema: %w", autoloadParamConfigPath, err)
+	}
+
+	pCfg.PipelineLogLevel, err = log.ParseLevel(pCfg.pipelineLogLevel)
+	if err != nil {
+		// Belt and suspenders.  Valid() should have caught this
+		return nil, fmt.Errorf("MakePipelineConfig(): config file (%s) had mal-formed log level: %w", autoloadParamConfigPath, err)
+	}
+
+	return &pCfg, nil
+
+}
+
+// Pipeline is a struct that orchestrates the entire
+// sequence of events, taking in importers, processors and
+// exporters and generating the result
+type Pipeline interface {
+	Start() error
+	Stop() error
+}
+
+type pipelineImpl struct {
+	ctx    context.Context
+	cfg    *PipelineConfig
+	logger *log.Logger
+
+	initProvider *data.InitProvider
+
+	importer   *importers.Importer
+	processors []*processors.Processor
+	exporter   *exporters.Exporter
+}
+
+func (p *pipelineImpl) Start() error {
+	p.logger.Infof("Starting Pipeline Initialization")
+
+	// TODO Need to change interfaces to accept config of map[string]interface{}
+
+	importerLogger := log.New()
+	importerLogger.SetFormatter(
+		PluginLogFormatter{
+			Formatter: &log.JSONFormatter{
+				DisableHTMLEscape: true,
+			},
+			Type: "Importer",
+			Name: (*p.importer).Metadata().Name(),
+		},
+	)
+
+	// TODO modify/fix ?
+	jsonEncode := string(json.Encode(p.cfg.Importer.Config))
+	genesis, err := (*p.importer).Init(p.ctx, plugins.PluginConfig(jsonEncode), importerLogger)
+
+	currentRound := basics.Round(0)
+
+	var initProvider data.InitProvider = &PipelineInitProvider{
+		currentRound: &currentRound,
+		genesis:      genesis,
+	}
+
+	p.initProvider = &initProvider
```

Comment:
```suggestion
	p.initProvider = &PipelineInitProvider{
		currentRound: &currentRound,
		genesis:      genesis,
	}
```

Also, consider renaming to `pluginProvider` if we're using the same one for all plugin types.